### PR TITLE
[Serve] Propagate replica constructor error to deployment status message and print num retries left

### PR DESCRIPTION
**Pull request by akyang-anyscale**
## Why are these changes needed?

This change will surface the replica constructor error as soon as the replica constructor fails for whatever reason. The exception will be populated in the deployment status so that it's viewable from the ray dashboard. Additionally, the number of replica constructor retries left will also be updated in the error message. This will help users more quickly debug a deployment that is failing to start.

## Related issue number

Closes https://github.com/ray-project/ray/issues/35604

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/deployment_state.py [hunk 1]
```diff
@@ -1942,6 +1944,22 @@ def check_curr_status(self) -> Tuple[bool, bool]:
                 )
                 return False, any_replicas_recovering
 
+        if failed_to_start_count > self._prev_replica_constructor_retry_counter:
```
Comment (zcin) [inline python/ray/serve/_private/deployment_state.py:1947]: Why is `_prev_replica_constructor_retry_counter` necessary? Can we just always call `update_message` and change it to a no-op if the message is the same?

File: python/ray/serve/tests/test_cli.py [hunk 2]
```diff
@@ -606,6 +606,34 @@ def check_for_failed_deployment():
...
+        status = yaml.safe_load(cli_output)["applications"][SERVE_DEFAULT_APP_NAME]
+        assert status["status"] == "DEPLOYING"
+
+        deployment_status = status["deployments"]["A"]
+        assert deployment_status["status"] == "UPDATING"
+        assert deployment_status["status_trigger"] == "CONFIG_UPDATE_STARTED"
+        assert "ZeroDivisionError" in deployment_status["message"]
+        return True
+
+    wait_for_condition(check_for_failed_deployment)
```
Comment (zcin) [inline python/ray/serve/tests/test_cli.py:634]: This seems to check for the status eventually showing up, which it should with current behavior when the replica fails three times. Does this test past without the changes in your PR?

**Review by zcin — COMMENTED** [review]
Overall LGTM, some questions!

File: python/ray/serve/_private/deployment_state.py (hunk 1 above)
Comment (akyang-anyscale) [reply python/ray/serve/_private/deployment_state.py:1947]: yeah you're right

File: python/ray/serve/tests/test_cli.py (hunk 2 above)
Comment (akyang-anyscale) [reply python/ray/serve/tests/test_cli.py:634]: It checks for the error message to show up when the deployment is still in the "DEPLOYING" phase to test the error is being propagated earlier. It does fail without the changes in the PR.

Comment (zcin) [reply python/ray/serve/tests/test_cli.py:634]: Ah gotcha, missed that! lgtm then

File: python/ray/serve/_private/deployment_state.py [hunk 3]
```diff
@@ -1942,6 +1942,22 @@ def check_curr_status(self) -> Tuple[bool, bool]:
...
+        else:
+            remaining_retries = failed_to_start_threshold - failed_to_start_count
+            message = (
+                f"A replica failed to start with exception. Retrying "
+                f"{remaining_retries} more time(s). Error:\n"
+                f"{self._replica_constructor_error_msg}"
+            )
+
+        if message != self._curr_status_info.message:
+            self._curr_status_info = self._curr_status_info.update_message(message)
```
Comment (zcin) [inline python/ray/serve/_private/deployment_state.py:1959]: might be easier to move the check into `update_message` method

**Review by zcin — APPROVED** [review]
LGTM

**Pull request merged**